File: src/api/endpoints/chatMessage/ChatMessageEndpoint.ts
import { Logger } from "ts-log";

import { HeaderProvider } from "../../../Func";
import ChatMessage from "../../../models/ChatMessage";
import ApiResponse from "../../ApiResponse";
import PaginatedResponse from "../../PaginatedResponse";
import ApiEndpoint from "../ApiEndpoint";
import TokenManager from "../auth/TokenManager";
import DefaultApiConfig from "../DefaultApiConfig";
import ChatMessageRequest from "./request/ChatMessageRequest";

/**
 * Endpoints for interacting with chat messages.
 * */
export default class ChatMessageEndpoint extends ApiEndpoint {
  constructor(config: DefaultApiConfig, tokenManager: TokenManager, log?: Logger, headerProvider?: HeaderProvider) {
    super(config, tokenManager, log, headerProvider);
  }

  protected override endpoint(): string {
    return "chatMessage";
  }

  /**
   * Fetch a chat message
   *
   * @param id - The id of the message
   * @returns The chat message
   *
   * */
  public async getMessage(id: string): Promise<ApiResponse<ChatMessage>> {
    return await this.getRequest<ChatMessage>(id);
  }

  /**
   * Gets the specified amount of chat messages for a given emergency.
   *
   * @param emergencyId - The emergency for which to fetch the chat history
   * @param limit - The number of emergencies to get, max 100
   * @param paginationToken - The string to use for pagination
   * */
  public async getMessageHistory(
    emergencyId: string,
    limit: number,
    paginationToken?: string,
  ): Promise<ApiResponse<PaginatedResponse<ChatMessage>>> {
    return await this.getRequest<PaginatedResponse<ChatMessage>>(`/conversation/${emergencyId}`, {
      limit,
      paginationToken,
    });
  }

  /**
   * Sends a new chat message
   *
   * @param message - The message to send
   * @returns The newly-created chat message
   *
   * */
  public async sendMessage(message: ChatMessageRequest): Promise<ApiResponse<ChatMessage>> {
    return await this.postRequest<ChatMessage>("", message);
  }

  /**
   * Update a chat message
   *
   * @param id - The id of the message to update
   * @param contents - The new content of the message
   * @returns The updated chat message
   *
   * */
  public async updateMessage(id: string, contents: string): Promise<ApiResponse<ChatMessage>> {
    return await this.putRequest<ChatMessage>(id, { contents });
  }

  /**
   * Delete a chat message
   *
   * @param id - The id of the message to delete
   *
   * */
  public async deleteMessage(id: string): Promise<ApiResponse> {
    return await this.deleteRequest(id);
  }
}

File: src-dotnet/api/endpoints/chatMessage/ChatMessageEndpoint.cs
namespace MedrunnerApiClient.Api.Endpoints.ChatMessage
{
    /// <summary>
    /// Endpoints for interacting with chat messages.
    /// </summary>
    public class ChatMessageEndpoint : ApiEndpoint
    {
        public ChatMessageEndpoint(ApiConfig config, TokenManager tokenManager, ILogger logger)
            : base(config, tokenManager, logger)
        {
        }

        protected override string Endpoint => "chatMessage";

        /// <summary>
        /// Fetch a chat message by id.
        /// </summary>
        public async Task<ApiResponse<Models.ChatMessage>> GetMessageAsync(string id)
        {
            return await GetRequestAsync<Models.ChatMessage>(id);
        }

        /// <summary>
        /// Gets the specified amount of chat messages for a given emergency.
        /// </summary>
        public async Task<ApiResponse<Api.PaginatedResponse<Models.ChatMessage>>> GetMessageHistoryAsync(string emergencyId, int limit, string? paginationToken = null)
        {
            var queryParams = new Dictionary<string, object> { { "limit", limit } };
            if (!string.IsNullOrEmpty(paginationToken))
                queryParams["paginationToken"] = paginationToken;
            return await GetRequestAsync<Api.PaginatedResponse<Models.ChatMessage>>($"/conversation/{emergencyId}", queryParams);
        }

        /// <summary>
        /// Sends a new chat message.
        /// </summary>
        public async Task<ApiResponse<Models.ChatMessage>> SendMessageAsync(object message)
        {
            return await PostRequestAsync<Models.ChatMessage>("", message);
        }

        /// <summary>
        /// Update a chat message.
        /// </summary>
        public async Task<ApiResponse<Models.ChatMessage>> UpdateMessageAsync(string id, string contents)
        {
            var payload = new { contents };
            return await PutRequestAsync<Models.ChatMessage>(id, payload);
        }

        /// <summary>
        /// Delete a chat message.
        /// </summary>
        public async Task<ApiResponse<string>> DeleteMessageAsync(string id)
        {
            return await DeleteRequestAsync<string>(id);
        }
    }
}
